clients/src/main/java/org/apache/kafka/common/utils/LogContext.java

```diff
@@ -40,20 +42,420 @@ public LogContext() {
+                logger.log(marker, fqcn, LocationAwareLogger.TRACE_INT, addPrefix(formattedMsg), new Object[]{arg1, arg2}, null);
+            }
+        }
+
+        @Override
+        public void trace(Marker marker, String format, Object... argArray) {
+            if (logger.isTraceEnabled()) {
+                String formattedMsg = MessageFormatter.arrayFormat(format, argArray).getMessage();
+                logger.log(marker, fqcn, LocationAwareLogger.TRACE_INT, addPrefix(formattedMsg), argArray, null);
+            }
+        }
+
+        @Override
+        public void trace(Marker marker, String msg, Throwable t) {
+            if (logger.isTraceEnabled()) {
+                logger.log(marker, fqcn, LocationAwareLogger.TRACE_INT, addPrefix(msg), null, t);
+
+            }
+        }
+
+        @Override
+        public void debug(String message) {
+            if (logger.isDebugEnabled()) {
+                logger.log(null, fqcn, LocationAwareLogger.DEBUG_INT, addPrefix(message), null, null);
+            }
+        }
+
+        @Override
+        public void debug(String format, Object arg) {
+            if (logger.isDebugEnabled()) {
+                final String formattedMsg = MessageFormatter.format(format, arg).getMessage();
+                logger.log(null, fqcn, LocationAwareLogger.DEBUG_INT, addPrefix(formattedMsg), new Object[]{arg}, null);
+            }
+        }
+
+        @Override
+        public void debug(String format, Object arg1, Object arg2) {
+            if (logger.isDebugEnabled()) {
+                final String formattedMsg = MessageFormatter.format(format, arg1, arg2).getMessage();
+                logger.log(null, fqcn, LocationAwareLogger.DEBUG_INT, addPrefix(formattedMsg), new Object[]{arg1, arg2}, null);
+            }
+        }
+
+        @Override
+        public void debug(String format, Object... args) {
+            if (logger.isDebugEnabled()) {
+                String formattedMsg = MessageFormatter.arrayFormat(format, args).getMessage();
+                logger.log(null, fqcn, LocationAwareLogger.DEBUG_INT, addPrefix(formattedMsg), args, null);
+            }
+        }
+
+        @Override
+        public void debug(String msg, Throwable t) {
+            if (logger.isDebugEnabled()) {
+                logger.log(null, fqcn, LocationAwareLogger.DEBUG_INT, addPrefix(msg), null, t);
+            }
+        }
+
+        @Override
+        public void debug(Marker marker, String msg) {
+            if (logger.isDebugEnabled()) {
+                logger.log(marker, fqcn, LocationAwareLogger.DEBUG_INT, addPrefix(msg), null, null);
+            }
+        }
+
+        @Override
+        public void debug(Marker marker, String format, Object arg) {
+            if (logger.isDebugEnabled()) {
+                String formattedMsg = MessageFormatter.format(format, arg).getMessage();
+                logger.log(marker, fqcn, LocationAwareLogger.DEBUG_INT, addPrefix(formattedMsg), new Object[]{arg}, null);
```

Comment:
I'm puzzling a little over the use of the arg array in these calls given that we've already used the argument to format the message. In the Log4j implementation, I see that this parameter is ignored, so I wonder if we can just always pass null?

Reply:
Thanks for the update. 

I am not sure about passing null here because some implementations might use the arg array. For example in 'logback' if I want to log an exception with two or more format arguments like below:
        logger.error("User with id {} not found in db {}", "a", "b", new RuntimeException("some exception"));
And the arg array is passed as null then the logger misses the exception:
[main] ERROR com.fico.ngc.FlUser - User with id a not found in db b


But if we pass arg array as is, then logger is able to capture the exception trace properly and prints as shown below:
00:29:52.210 [main] ERROR com.fico.ngc.FlUser - User with id a not found in db b
java.lang.RuntimeException: some exception
	at com.fico.ngc.FlUser.main(FlUser.java:6)

Reply:
Thanks for the explanation. I checked the logback implementation and it does indeed use the argument array. Maybe it's not such a big deal since we have the log level guard.